### ISSUE TITLE
"Features" - a new API for Archi

### ISSUE THREAD

**Issue by Phillipus** [archimatetool/archi#530]
Since earlier this year I've been working on a new API called "Features".

Here's the problem it tries to solve (at least partially):

Every time we add a new feature attribute to the Ecore model, such as opacity, border, we have to regenerate the Ecore model files and bump up the ModelVersion number so we can check for backwards compatibility. This means that model files are not always compatible and you get a warning dialog box. If you do open the model in an older version of Archi and re-save it you will lose the ecore attribute or element.

"Features" works like the existing "Properties" but with added functionality and easier to use. Once we have a version of Archi with this in any new feature will be preserved even if the older version of Archi cannot do anything with it. It will be ignored, but will persist on save.

A feature is a name/value pair that is not intended to be directly exposed in the UI except for how it it is used (like an Ecore attribute or element).

Every object can have features - the model, a concept, a folder, a View, a diagram component.

An example - a boolean flag to hide a connection label. With "Features" you do this:

```java
IFeatures features = connection.getFeatures();
features.putBoolean("connectionLabelVisible", false);
```

And the XML is saved like this:

```
<sourceConnection>
   <feature name="connectionLabelVisible" value="false"/>
</sourceConnection>
```

The Features framework can put and get strings, ints, booleans as a convenience. Each value is ultimately a string when persisted.

Each feature is scoped to its parent object.

This means that third-party plugins can persist their data in features without having to expose it in the User Properties table, although they can expose it in other ways if they wish. Also, in coArchi we could expose the features in a table to show diffs when merging.

Things I haven't resolved:

- A feature, once persisted in the *.archimate file will always be there. I could set a default value for the feature and if the value contains the default value it is not written (like Ecore does it). Downside is that when looking at the *archimate file you have no idea if a feature is present if it is set to the default value.
- If we like this idea, when to add it to Archi?

There is a branch called "features" if you want to take a look (this branch is constantly rebased and cherry picked as I work on it)

**Comment by herve91** [comment]
These "features" would be very convenient for my plugins as they would allow to persist data without exposing them into properties. Allowing to store images the same way (without having to expose them in a view) would also be very convenient :-)

**Comment by Phillipus** [comment]
TL;DR - Features are like Properties but hidden from the user and have the same functionality as eCore attribute features.

**Comment by Phillipus** [comment]
> A feature, once persisted in the *.archimate file will always be there. I could set a default value for the feature and if the value contains the default value it is not written (like Ecore does it). Downside is that when looking at the *archimate file you have no idea if a feature is present if it is set to the default value.

I've added this as an option.

**Comment by Phillipus** [comment]
@jbsarrodie What do you think about the features feature? Is it a feature that you'd like to see featured? :-)

**Comment by jbsarrodie** [comment]
> @jbsarrodie What do you think about the features feature? Is it a feature that you'd like to see featured? :-)

That's a nobrainer for me: this is absolutly needed to ease future developments.

Of course this means that this has to be exposed through jArchi, and we still have to decide what to do with already existing attributes stored as ECore attributes? I'd suggest to migrate all of them to feature at some point.

**Comment by Phillipus** [comment]
> Of course this means that this has to be exposed through jArchi,

It will be much like the API we have for Properties. And it would need also to be added to coArchi for the diff viewer.

> I'd suggest to migrate all of them to feature at some point.

Yes, perhaps. Once the features feature has been established for a while.

Any comments about the Features API? As far as I can see a name/value pair is sufficient, as this is basically what an eCore attribute is. Third parties can namespace their entries with a unique prefix like put("org.myplugin.killerfeature", "awesomeValues")

Shall we aim for 4.6? We could add it but not use it for any new feature in that build.

**Comment by Phillipus** [comment]
In fact, thinking about it, if we only had Features it would make updating jArchi so much easier. We wouldn't need specific getters and setters.

**Comment by Phillipus** [comment]
When we add Feature we can remove or deprecate the root Metadata element. Does anyone use it?

**Comment by jbsarrodie** [comment]
> Does anyone use it?

At least not me

**Comment by Phillipus** [comment]
I know that @herve91 was going to use it in the Specialization plug-in. Perhaps use the Feature instead?

**Comment by Phillipus** [comment]
I guess the most obvious use for it right now is for outline opacity (alpha) for diagram objects. I've already implemented this as an Ecore attribute but we haven't released it yet (for 4.6). I could convert it to a Feature for 4.6 instead.

**Comment by Phillipus** [comment]
As I work on re-implementing the outline alpha as an Archi "feature" I find that the new API is confusing because of its name. Ecore also uses "Feature", So I'm doing things like getArchiFeature(). 

Perhaps we should have a different name than "feature". (But not "attribute")

Facet?

**Comment by Phillipus** [comment]
> Facet?

No, that's used in XSD as a restriction.

Factor?
Aspect?
Detail?

**Comment by Phillipus** [comment]
Wait....I have a better idea. I can convert refactor existing code that refers to Ecore features to refer to them by name and also Archi feature by name. So "feature" actually has advantages.

**Comment by jbsarrodie** [comment]
>But not "attribute"

We have to keep in mind that what Archi call Properties are in fact Attributes for the ArchiMate standard. And so, at some point (major version ?) we should switch and use "Attributes" (at least is UI and jArchi).

**Comment by jbsarrodie** [comment]
BTW, we must be able to add such 'feature' to concepts but also diagram nodes (and in fact every where).

**Comment by Phillipus** [comment]
> We have to keep in mind that what Archi call Properties are in fact Attributes for the ArchiMate standard. And so, at some point (major version ?) we should switch and use "Attributes" (at least is UI and jArchi).

Yes, I already made some plans for that for later.

> BTW, we must be able to add such 'feature' to concepts but also diagram nodes (and in fact every where).

Already everywhere. :-)

> Every object can have features - the model, a concept, a folder, a View, a diagram component.

**Comment by Phillipus** [comment]
> I guess the most obvious use for it right now is for outline opacity (alpha) for diagram objects. I've already implemented this as an Ecore attribute but we haven't released it yet (for 4.6). I could convert it to a Feature for 4.6 instead.

Done, and it works.

**Comment by Phillipus** [comment]
@jbsarrodie I also have an experimental feature that hides/shows the name label for a connection in a View. Is it something that would be useful?

**Comment by jbsarrodie** [comment]
> @jbsarrodie I also have an experimental feature that hides/shows the name label for a connection in a View. Is it something that would be useful?

Yes, that would be helpful in a number of situations.

**Comment by Phillipus** [comment]
> Yes, that would be helpful in a number of situations.

OK, I'll get that in for 4.6.

I've put it in the "Appearance" tab:

![Image 1](https://user-images.githubusercontent.com/600504/65372662-e77ec600-dc6a-11e9-8979-eade785aaf1f.png)

**Comment by jbsarrodie** [comment]
I'd move 'Show name' at the top and (if possible) 'Text position', 'Font' and 'Font color' should be gray out when show name is disabled.

**Comment by Phillipus** [comment]
"Features" has been merged to master branch.

**Comment by Phillipus** [comment]
I've marked IMetadata getters and setters as deprecated.

@herve91 Will you be able to use the Features instead of Metadata?

**Comment by herve91** [comment]
I will have to :-)

But I need to hurry, I haven't had time to adapt all my plugins to 4.5, I cannot miss 4.6 ...

**Comment by Phillipus** [comment]
Closing. If any issues are found with Features in 4.6. please open new issues.
